New Extension: Particle Engine

SharkPool-SP opened the pull request:
# Create Powerful Particle Engines with NO Clones
https://github.com/user-attachments/assets/459d44a6-9f02-4d2a-97c5-80be29173808

https://github.com/user-attachments/assets/8d0669f7-7c75-4ab7-9bbd-71c6df2a309d

SharkPool-SP committed:
Create Particle-Engine.js

SharkPool-SP committed:
bye bye lint

SharkPool-SP committed:
Create Particle-Engine.svg

BludIsAnLemon commented:
Peak!!

Brackets-Coder commented:
Two questions I have: 

1. Does high quality pen affect rendering quality?
2. I've noticed that a lot of your extensions have gradient blocks. This looks cool, but the way you're implementing it with the patch is very weird. I think it'd be sick if we could use CSS colors and gradients as the color values for blocks instead of just hex values. How involved in the scratch-gui would it be to be able to do this?

theapplesguy2 commented:
yes

hammouda101010 commented:
WOW! a new particle engine with no clones! that's great for performance and now overflow your projects with clones! (even if i use clones, a lot)

SharkPool-SP commented:
> Two questions I have: 
> 
> 
> 
> 1. Does high quality pen affect rendering quality?
> 
> 2. I've noticed that a lot of your extensions have gradient blocks. This looks cool, but the way you're implementing it with the patch is very weird. I think it'd be sick if we could use CSS colors and gradients as the color values for blocks instead of just hex values. How involved in the scratch-gui would it be to be able to do this?

1. no, this is not a "pen" canvas, to change quality, you'd need to use the set stage size block, but it comes at a cost of lower performance.

2. The gui doesn't allow gradients in blocks at all. So I can't simply just put it in the extension block code. So I have to use the patch.

Additionally, blocks in the editor are all SVGs, which means you can't color them with any html property. You have do what I do and make SVG gradients or patterns.

hammouda101010 commented:
> Two questions I have:
> 
> 1. Does high quality pen affect rendering quality?
> 2. I've noticed that a lot of your extensions have gradient blocks. This looks cool, but the way you're implementing it with the patch is very weird. I think it'd be sick if we could use CSS colors and gradients as the color values for blocks instead of just hex values. How involved in the scratch-gui would it be to be able to do this?

well, i just ~~stole~~ borrowed code for you to use the patch:

function add2Body() {
    var svg = document.createElement("div");
    svg.innerHTML = `<svg><defs>
      <linearGradient x1="100" y1="0" x2="100" y2="200" id="SPpartEngine-GRAD1" gradientUnits="userSpaceOnUse"><stop offset="0" stop-color="#0090ff"></stop><stop offset="50%" stop-color="#0000ff"></stop></linearGradient>
      </defs></svg>`;
    document.body.appendChild(svg);
  }
  if (Scratch.gui) Scratch.gui.getBlockly().then((SB) => {
    add2Body();
    if (!SB?.SPgradients?.patched) { // Gradient Patch by 0znzw & SharkPool
      SB.SPgradients = {gradientUrls: {}, patched: false};
      const BSP = SB.BlockSvg.prototype, BSPR = BSP.render;
      BSP.render = function(...args) {
        const res = BSPR.apply(this, args);
        let category;
        if (this?.svgPath_ && this?.category_ && (category = this.type.slice(0, this.type.indexOf("_"))) && SB.SPgradients.gradientUrls[category]) {
          const urls = SB.SPgradients.gradientUrls[category];
          if (urls) this.svgPath_.setAttribute("fill", urls[0]);
        }
        return res;
      }
      SB.SPgradients.patched = true;
    }
    SB.SPgradients.gradientUrls["SPpartEngine"] = ["url(#SPpartEngine-GRAD1)"];
  });

Brackets-Coder commented:
> > Two questions I have:
> > 
> > 1. Does high quality pen affect rendering quality?
> > 2. I've noticed that a lot of your extensions have gradient blocks. This looks cool, but the way you're implementing it with the patch is very weird. I think it'd be sick if we could use CSS colors and gradients as the color values for blocks instead of just hex values. How involved in the scratch-gui would it be to be able to do this?
> 
> well, i just ~stole~ borrowed code for you to use the patch:
> 
> function add2Body() { var svg = document.createElement("div"); svg.innerHTML = `<svg><defs> <linearGradient x1="100" y1="0" x2="100" y2="200" id="SPpartEngine-GRAD1" gradientUnits="userSpaceOnUse"><stop offset="0" stop-color="#0090ff"></stop><stop offset="50%" stop-color="#0000ff"></stop></linearGradient> </defs></svg>`; document.body.appendChild(svg); } if (Scratch.gui) Scratch.gui.getBlockly().then((SB) => { add2Body(); if (!SB?.SPgradients?.patched) { // Gradient Patch by 0znzw & SharkPool SB.SPgradients = {gradientUrls: {}, patched: false}; const BSP = SB.BlockSvg.prototype, BSPR = BSP.render; BSP.render = function(...args) { const res = BSPR.apply(this, args); let category; if (this?.svgPath_ && this?.category_ && (category = this.type.slice(0, this.type.indexOf("_"))) && SB.SPgradients.gradientUrls[category]) { const urls = SB.SPgradients.gradientUrls[category]; if (urls) this.svgPath_.setAttribute("fill", urls[0]); } return res; } SB.SPgradients.patched = true; } SB.SPgradients.gradientUrls["SPpartEngine"] = ["url(#SPpartEngine-GRAD1)"]; });

I've already looked at the code, but I don't intend to use it for my extensions. Thank you anyway though.

Brackets-Coder commented:
> > Two questions I have:
> > 
> > 1. Does high quality pen affect rendering quality?
> > 2. I've noticed that a lot of your extensions have gradient blocks. This looks cool, but the way you're implementing it with the patch is very weird. I think it'd be sick if we could use CSS colors and gradients as the color values for blocks instead of just hex values. How involved in the scratch-gui would it be to be able to do this?
> 
> 1. no, this is not a "pen" canvas, to change quality, you'd need to use the set stage size block, but it comes at a cost of lower performance.
> 2. The gui doesn't allow gradients in blocks at all. So I can't simply just put it in the extension block code. So I have to use the patch.
> 
> Additionally, blocks in the editor are all SVGs, which means you can't color them with any html property. You have do what I do and make SVG gradients or patterns.

Interesting. I was just curious as to if we could change the scratch-gui repo to add some sort of functionality for gradients and CSS colors in the block color APIs, but I guess not - It didn't occur to me that the SVGs can't use CSS gradients. Maybe we could use this patch though, and manually inject the colors into the patch? IDK, it's just an idea. Probably won't be done though.

SharkPool-SP committed:
Particle-Engine -- Interpolation

SharkPool-SP commented:
Theres now an interpolation option! :)

hammouda101010 reviewed (approved):
ok looks cool

SharkPool-SP committed:
Particle-Engine -- Waves and Exports

SharkPool-SP commented:
!format

SharkPool-SP committed:
Particle-Engine -- Improve Tinting

SharkPool-SP committed:
Particle-Engine.js -- new gradient patch

SharkPool-SP committed:
Particle-Engine.js -- small fixes

SharkPool-SP committed:
Particle-Engine.js -- WebGL Rewrite + Improvements

CubesterYT commented:
!format

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
Particle-Engine.js -- lint shut

SharkPool-SP committed:
Particle-Engine.js -- LINT SHUT

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
unused log

SharkPool-SP committed:
add a cool trails option

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
Particle-Engine.js -- use WebgGL2 if availiable

SharkPool-SP committed:
Particle-Engine.js -- Blend Modes, Fixes, and Improvements

SharkPool-SP commented:
!format

SharkPool-SP commented:
blending blocks!
![image](https://github.com/user-attachments/assets/7e00b54a-afc2-41fb-95e9-5a091b106c8f)

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
Particle-Engine.js -- Gradient Patch V3

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
Particle-Engine.js -- do not draw if off-screen

DangoCat committed:
[Automated] Format code

LSPECTRONIZTAR commented:
No offense but I genuinely don't think this will ever get merged
Still a peak extension, though

SharkPool-SP committed:
Particle-Engine.js -- gradient patch update

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
Particle-Engine.js -- optimize

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

Brackets-Coder reviewed (changes requested):
Seems to work well for me. This extension is peak, but here are some things to consider:
1. It runs even while the project is stopped, this can be confusing to some beginners who are likely to use this extension since it's so helpful.
2. It's fairly easy to pick up, but have you considered writing documentation?
3. The banner looks good but has text

SharkPool-SP commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1350 @@
+              TARGET: { type: Scratch.ArgumentType.STRING, menu: "TARGETS" },
+            },
+          },
+          {
+            blockType: Scratch.BlockType.LABEL,
+            text: Scratch.translate("Advanced"),
+          },
```
yea cuz its more advanced

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1350 @@
+  const Cast = Scratch.Cast;
+  const runtime = vm.runtime;
+  const render = vm.renderer;
+  const { Drawable, Skin, twgl } = render.exports;
```
> it literally says right there where im getting those from. renderer.exports.
> 
> Look in the developer console on turbowarp and you'll see them when you put the code in

I see that but I didn't know renderer.exports existed

SharkPool-SP replied:
well er it does :trollface:

SharkPool-SP commented:
> > > Basically just remove the text from the banner then I guess we're good to merge this peak
> 
> > 
> 
> > I think documentation can be saved for another time, the blocks are in order of usability. So it _should_ be easy to pickup. It works closely to the geometry dash particle engine as well.
> 
> > 
> 
> > We can always go back and add it if need be
> 
> 
> 
> In my testing it didn't take me long to figure out but I couldn't get engines working on sprites, only the stage? Maybe I was doing something wrong

Wdym. The engines don't follow sprites, they're just attached to them

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1350 @@
+// Name: Particle Engine
+// ID: SPpartEngine
+// Description: Create powerful Particle Engines without Clones
```
```suggestion
// Description: Create powerful Particle Engines without Clones.
```

Brackets-Coder committed:
Update extensions/SharkPool/Particle-Engine.js

add punctuation to end of description comment

SharkPool-SP commented:
> > Wdym. The engines don't follow sprites, they're just attached to them
> 
> 
> 
> No, I meant I for some reason couldn't get them working when attached to a sprite, I rushed through it so I must've done something wrong

Maybe you didn't make an emitter for that engine?

Brackets-Coder commented:
> Maybe you didn't make an emitter for that engine?

Oops I got it working sorry

Brackets-Coder commented:
> Is that a gradient for the block colors? Looks really nice.

Yes

PPPDUD commented:
> > > > Is that a gradient for the block colors? Looks really nice.
> > > 
> > > 
> > > Yes
> > 
> > 
> > I didn't know that block colors could use _gradients_!
> 
> They can't by default, it's an SVG patch.

I see.

SharkPool-SP committed:
remove svg text

SharkPool-SP commented:
alright fixed

Brackets-Coder reviewed (dismissed):
Thanks, everything looks good!

Brackets-Coder commented:
@PPPDUD

PPPDUD commented:
> @PPPDUD

I'm going to have to pass for tonight, but I might be willing to review tomorrow.

SharkPool-SP commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1350 @@
+// Name: Particle Engine
+// ID: SPpartEngine
+// Description: Create powerful Particle Engines without Clones.
```
its already in senence case??

Brackets-Coder committed:
fix sentence case

GarboMuffin reviewed (commented):
- when i change the VM fps to 60, i would expect the particles to not move twice as fast. there does seem to be delta time logic at least in some places. is it not being used? ✅
- when i have an emitter going and i resize the stage, i would expect the particle canvas to resize too and at least appear in the right spot after the resize (as is, stage emitter is very offset to the right after resizing 480x360 to 640x360) ✅
- the dropdowns in the flyout only update after exiting the editor and going back in; newly creates sprites will not show up ✅ 
- uploading one webgl context to another context seems to be slow. create like 10 particle engines, even with almost nothing going on, it gets really slow: https://share.firefox.dev/4gty2bI ✅ **[Not Fixing, other gallery extensions use this method]**
- chrome limits max webgl contexts per tab to something like 16, so by the time someone makes their 16th particle context, the context used by the vm will be lost until a full refresh ✅
- even if unused engines are removed with the block, it's not guaranteed that chrome will garbage collect them and release the webgl slot right away (make a "create engine for x / remove engine for x" script and duplicate it on top of itself 20 times to see example)
- blend modes are seemingly affected by same problems as in https://github.com/TurboWarp/extensions/pull/2142 (set particle blend mode to subtract in light mode for an example) ✅
- the particles are highly pixelated in fullscreen and on high dpi displays outside of fullscreen ✅ **[Not Fixing]**

Brackets-Coder commented:
Next time I'll review more thoroughly...

SharkPool-SP commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1350 @@
+  }
+
+  // Util Funcs
+  const fastSin = (rads) => sinTable[Math.floor(rads * INV_TWO_PI) & 8191];
+  const fastCos = (rads) => cosTable[Math.floor(rads * INV_TWO_PI) & 8191];
```
iirc I was able to squeeze +100 particle emission

SharkPool-SP commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1350 @@
+    }
+  }
+
+  if (Scratch.gui)
```
and Id rather not wait 6 months just to get a PR merged from you

SharkPool-SP commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1350 @@
+      const target = this.getSprite(args.TARGET);
+      if (target && target[engineTag]) {
+        const engine = target[engineTag];
+        if (engine.emitters[args.NAME])
```
emitters already is `emitters: Object.create(null)`. Look at the makeEngine function

SharkPool-SP committed:
Particle-Engine.js -- fix some garbo issues (1/2)

SharkPool-SP committed:
fix more garbo stuff (2/3)

SharkPool-SP committed:
fix garbo issues 3/3

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
Fix emitter position on canvas size change

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

Brackets-Coder commented:
Remind me to review again this Saturday, please

SharkPool-SP commented:
> Remind me to review again this Saturday, please

Saturday

Brackets-Coder commented:
> > Remind me to review again this Saturday, please
> 
> Saturday

You may have to wait until tomorrow, today was busier than I expected and I think I overbooked on extension reviews lol

SharkPool-SP commented:
Tomorrow

Brackets-Coder commented:
> Tomorrow

Ha ha hilarious, I'll take a look

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1408 @@
+    const canvas = document.createElement("canvas");
+    const gl = canvas.getContext("webgl2") ?? canvas.getContext("webgl");
+    canvas.width = runtime.stageWidth || 480;
+    canvas.height = runtime.stageHeight || 360;
```
I have a few concerns about the potential performance limitations of creating a new canvas for each engine. Running multiple engines simultaneously on an older or lower end machine may eat up power, but I haven't run into issues with my minimalistic testing on my computer

SharkPool-SP replied:
im assuming I can resolve this comment as you have another below that talks about this specifically

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1408 @@
+      canvas.height = size[1];
+      engine.projection = twgl.m4.ortho(0, size[0], size[1], 0, -1, 1);
+    }).bind(this);
+    vm.on("STAGE_SIZE_CHANGED", drawable.stageSZChange);
```
I don't know if it should be here, but have you considered adding high quality pen support? I tried changing the "stage size engine" block, no higher resolution. Tried turning on HQPen, no increase in resolution. Might be nice to have

SharkPool-SP replied:
This doesnt increase resolution it increases the drawing boundaries.

You can manually increase the resolution by increasing the canvas size block (with the `set canvas size` block) and by drawing bigger sized particles

This is not pen related so I wont bother making it active on HQ Pen. This can be specified in documentation

Brackets-Coder replied:
Okay
just saying the way Simple3D pulled it off was kinda cool
I guess I was just too dumb to make it higher resolution

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1408 @@
+
+    checkEngineContexts() {
+      // its fair to trust that the user knows what theyre doing if they ingore this
+      if (engineCount > ENGINE_CAP) {
```
Ah - about the performance
I think you mentioned the default max number of engines was 10?
wait never mind checking it's 7
Obviously there's no block to control this, but again — I feel like this is something that should be clarified in documentation

SharkPool-SP replied:
will mention in documentation

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1408 @@
+    deleteEmitter(args) {
+      const target = this.getSprite(args.TARGET);
+      if (target && target[engineTag])
+        delete target[engineTag].emitters[args.NAME];
```
Like I said earlier, it might be nice if these weren't immediately deleted, but told to stop adding particles to the engine and then are officially deleted from the object when there are no particles being rendered. It's just a nice nitpick though, no need to worry

SharkPool-SP replied:
done

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1350 @@
+        const engine = target[engineTag];
+        const space = engine.emitters[args.NAME];
+        if (space) {
+          newTexture(args.IMG, engine.gl, (texture) => {
```
> this leaks arbitrary http requests without going through canFetch

May I inquire as to why this was unresolved with no explanation?

SharkPool-SP replied:
if you look at 'newTexture' it uses canFetch

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1408 @@
+        case "star":
+          return shapes.star;
+        default:
+          return "";
+      }
+    }
```
This can probably be streamlined if you rename some of the properties;
```suggestion
    defaultTexture(args) {
      return shapes[args.TYPE] || "";
    }
```
But I'm not complaining and it's probably too much of a hassle for only a few lines

SharkPool-SP replied:
gonna use a modified version since people can input special codes like 'toString' or 'constructor' on the object

Brackets-Coder replied:
Just an example

Brackets-Coder commented on images/SharkPool/Particle-Engine.svg:
Really nice banner

SharkPool-SP replied:
thanks

Brackets-Coder reviewed (changes requested):
Alright, this is another "quick" review that took me forever because I went on side quests. I haven't tested everything, but with a few exceptions, what I've looked at seems good!

- [x] I verified that the extension follows all acceptance criteria and guidelines in CONTRIBUTING.md.
 - [x] I believe the block list is well-designed and intuitive. _- yeah for the most part_
 - [x] I tested **~~all~~** **_most_** functionality and found no bugs, security flaws, or unintuitive behavior. 
 - [ ] I reviewed the code and believe it is well-written and easy to maintain or extend in the future, even without help from the original author. - I went through it, but didn't critically analyze it. I'm unfamiliar with some of the twgl and patches stuff
 - [ ] I have no additional concerns.

I don't know if this is a bad idea:
Creating an emitter has a nice transition where the particles begin to fill the scene. Deleting an emitter immediately makes all particles disappear, which is contrary to what I expected happening: stopping the emission process and waiting for all particles to fall away from the scene. Instead, all of the particles immediately disappeared, which I expected might happen if I remove the engine, not the emitter. This can be mitigated by setting the max particle count to 0 and waiting a bit before deleting the emitter, but either way I still think it might be a nice feature to have.

- Sometimes [y] position of emitter reports -0, which isn't technically incorrect, but it's still a bit odd to see, considering the x-position reporter doesn't do this

- Making an emitter on a sprite and then making the sprite draggable does not make the emitter follow the sprite. This makes sense to me since such functionality would be achieved with the "Set position" block, but it's something you may want to clarify in documentation.

- Interpolation seems to work really well, but you might want to add some kind of option that automatically enables that feature if Interpolation is turned on in runtime options

- It also seems like you fixed the delta time issue, nice job! I'm having a bit of issues with that myself, see [this comment](https://github.com/TurboWarp/extensions/pull/1974#issuecomment-3688244374). Let me know if you have any tricks you can tell me 😉 

Big takeaways: 
Documentation will clarify a lot of potentially confusing stuff for your users. 

Other than all of that, nice work!

SharkPool-SP committed:
requested changes

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

SharkPool-SP commented:
will write documentation in the morning, in the meantime @Brackets-Coder I fixed all your requests, including:
<img width="248" height="51" alt="Screenshot 2026-01-17 at 1 47 40 AM" src="https://github.com/user-attachments/assets/aa044fee-0561-40d6-9f1d-0095a75fe563" />

SharkPool-SP committed:
Add new extensions to extensions.json

SharkPool-SP committed:
Fix license spelling

Brackets-Coder commented:
I guess today is going to be a review day for me... I'll see what I can do

Brackets-Coder commented on extensions/extensions.json:
why are there other additions? Sync the branch maybe so the newest extensions are already there?

SharkPool-SP replied:
I don't trust GitHub to accurately merge it, regardless shouldn't affect the actual file when merged

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1455 @@
+      if (engineCount > ENGINE_CAP) {
+        const warningText = Scratch.translate(
+          "WARNING: Particle Engine -- reached unstable number of engines!"
+        );
```
This isn't necessary, but if you want an actual m-dash character instead of just two hyphens here it is 😉 :

```suggestion
        const warningText = Scratch.translate(
          "WARNING: Particle Engine — reached unstable number of engines!"
        );
```

It should display longer in the browser/editor

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1455 @@
+        rawPos: [0, 0],
+        opts: structuredClone(DEFAULT_BEHAVIOURS),
+        frameCnt: 0,
+        tintCache: new Map(),
+        data: new Set(),
+      };
+      newTexture(PRESET_SHAPES["square"], engine.gl, (t) => {
+        engine.emitters[args.NAME].texture = t;
+      });
+      runtime.requestRedraw();
+    }
+
+    deleteEmitter(args) {
+      const target = this.getSprite(args.TARGET);
+      if (target && target[engineTag]) {
+        delete target[engineTag].emitters[args.NAME];
+      }
+    }
+
+    deleteEmitterWait(args, util) {
+      let target, emitter;
+      if (util.stackFrame.emitterCallback) {
+        [target, emitter] = util.stackFrame.emitterCallback;
+        if (emitter.data.size > 0) {
+          util.yield();
+        } else {
+          delete target[engineTag].emitters[args.NAME];
+        }
+      } else {
+        target = this.getSprite(args.TARGET);
+        if (!target || !target[engineTag]) return;
+
+        emitter = target[engineTag].emitters[args.NAME];
+        if (!emitter) return;
+
+        emitter.opts.emission = { val: 0, inf: 0 };
+        util.stackFrame.emitterCallback = [target, emitter];
+        util.yield();
+      }
+    }
```
Nice.

Brackets-Coder commented on extensions/SharkPool/Particle-Engine.js:
```diff
@@ -0,0 +1,1455 @@
+      }
+    }
+
+    defaultTexture(args) {
+      return PRESET_SHAPES[Cast.toString(args.TYPE)] || "";
+    }
```
Definitely better

Brackets-Coder reviewed (changes requested):
One final concern before I approve

I like how interpolation can now be toggled both from the blocks and from runtime.
I like how sprite engines now go to the sprite and position offsets them.

Everything else looks as good as it did before

I think I might've discovered an issue where doing the following many times pulls up the "Too many engines" warning, even though I removed the engine

<img width="1336" height="592" alt="Image" src="https://github.com/user-attachments/assets/4b12bbb7-13d1-4261-bb82-d0b9aaa9c0c8" />

Also, the gradient patch appears to not export with the "export blocks as image" add-on...

SharkPool-SP committed:
documentation

SharkPool-SP committed:
Fix engine count not decrementing

SharkPool-SP commented:
Okay @Brackets-Coder fixed all those issues, added documentation, everything is done!

Again, the extensions.json having multiple additions is fine, its just copied from the actual json file. Additionally I cant fix the gradient patch not exporting svgs properly since thats an addon thing that I have no access to :/

Brackets-Coder commented:
> Additionally I cant fix the gradient patch not exporting svgs properly since thats an addon thing that I have no access to :/

I'm not concerned about that

SharkPool-SP commented:
> > Additionally I cant fix the gradient patch not exporting svgs properly since thats an addon thing that I have no access to :/
> 
> 
> 
> I'm not concerned about that

Okay, gimmie approval 🤤🤤

Brackets-Coder commented:
> Okay, gimmie approval 🤤🤤

Patience, young padawan. I'm going to go through and check again...

Brackets-Coder commented on docs/SharkPool/Particle-Engine.md:
````diff
@@ -0,0 +1,198 @@
+```scratch
+create engine for (Stage v) ::#2474ff
+```
+This block will initialize a new engine in the specified sprite.
````
It's unreasonable to ask to change every reference to also mention the stage, but I think this is fine how it is since the stage is in the dropdown
```suggestion
This block will initialize a new engine in the stage or the specified sprite.
```

Brackets-Coder commented on docs/SharkPool/Particle-Engine.md:
```diff
@@ -0,0 +1,198 @@
+
+These blocks will delete the specified emitter from the specified engine.
+
+The first block is instantaneous, meaning all currently emitted particles will disappear.
```
```suggestion
The first block is instantaneous, meaning all currently emitted particles will disappear immediately.
```

Brackets-Coder commented on docs/SharkPool/Particle-Engine.md:
```diff
@@ -0,0 +1,198 @@
+These blocks will delete the specified emitter from the specified engine.
+
+The first block is instantaneous, meaning all currently emitted particles will disappear.
+However the second block will wait until all particles disappear before deleting.
```
```suggestion
However, the second block will wait until all particles disappear before deleting.
```

Brackets-Coder reviewed (changes requested):
Removing all and creating a new engine multiple times still returns the error, but there's no popup: `There are too many active WebGL contexts on this page, the oldest context will be lost.` Even executing it multiple times should result in one final engine, not 7

e.g., +1 -1 +1 - 1 = 0, not 2

"remove (all)" block should delete every WebGL context

Additionally, the "delete and wait" block seems to not work with interpolation (it looks more stuttery)

SharkPool-SP commented:
> Removing all and creating a new engine multiple times still returns the error, but there's no popup: `There are too many active WebGL contexts on this page, the oldest context will be lost.` Even executing it multiple times should result in one final engine, not 7
> 
> e.g., +1 -1 +1 - 1 = 0, not 2
> 
> "remove (all)" block should delete every WebGL context
> 
> Additionally, the "delete and wait" block seems to not work with interpolation (it looks more stuttery)

Wdym the delete and wait thing, can you show a video??

SharkPool-SP committed:
docs suggestions

SharkPool-SP committed:
efficiently close engines

SharkPool-SP commented:
!format @Brackets-Coder fixed all those other issues

DangoCat committed:
[Automated] Format code

Brackets-Coder reviewed (changes requested):
- Turning on interpolation from runtime options does not enable interpolation in the particle system. Interpolation is only enabled via the block. It should be two-way
   - Turning interpolation on via block and then turning it off and on back again in runtime works
    - Turning interpolation off via block and then turning on via runtime options does not work
    - it's a bit complicated since turning interpolation off via block should not turn off runtime interpolation 

Here's the video for delete and wait:

https://github.com/user-attachments/assets/13a55a70-bca9-4af2-93cf-3c13b7e7822f

Brackets-Coder commented:
Sorry for being picky

Brackets-Coder commented:
Other than that you're good

SharkPool-SP commented:
> * Turning on interpolation from runtime options does not enable interpolation in the particle system. Interpolation is only enabled via the block. It should be two-way
>   
>   * Turning interpolation on via block and then turning it off and on back again in runtime works
>   * Turning interpolation off via block and then turning on via runtime options does not work
>   * it's a bit complicated since turning interpolation off via block should not turn off runtime interpolation
> 
> Here's the video for delete and wait:
> 
>  video.mp4

the problem with that is it wont be interpolation per engine anymore. What youre suggesting is that all engines should have interpolation enabled when runtime turns it on?

Brackets-Coder commented:
> the problem with that is it wont be interpolation per engine anymore. What youre suggesting is that all engines should have interpolation enabled when runtime turns it on?

It's complicated since there's two methods of toggle, one with the block and one without. Ideally both should "just work" intuitively but the two-way communication between both methods doesn't work "intuitively." To be honest I would just say keep either one and remove the other method, but if you can get both to work together that'd be better

SharkPool-SP committed:
Make interpolation automatic

SharkPool-SP committed:
docs change

Removed the interpolation option description from the Particle Engine documentation.

SharkPool-SP commented:
> > the problem with that is it wont be interpolation per engine anymore. What youre suggesting is that all engines should have interpolation enabled when runtime turns it on?
> 
> It's complicated since there's two methods of toggle, one with the block and one without. Ideally both should "just work" intuitively but the two-way communication between both methods doesn't work "intuitively." To be honest I would just say keep either one and remove the other method, but if you can get both to work together that'd be better

okay I just removed the option to toggle interpolation, I just made it automatically turn on

Brackets-Coder commented:
> okay I just removed the option to toggle interpolation, I just made it automatically turn on

Fantastic! If you're going to do it that way make sure you explain in the docs :D

Brackets-Coder commented:
I'll take a look later when I have time and probably approve

SharkPool-SP commented:
> I'll take a look later when I have time and probably approve

Already changed the docs

Brackets-Coder reviewed (changes requested):
Turning on interpolation and then creating an engine/emitter causes the emitter to be created without interpolation enabled.

Executing the "delete and wait" block still appears to disable interpolation on the emitter

Spamming the following block stack in the editor results in the `There are too many active WebGL contexts on this page, the oldest context will be lost` error.
<img width="744" height="304" alt="Image" src="https://github.com/user-attachments/assets/a0033eb8-27ff-4b70-b2ce-7037b77d9e93" />

It seems to be mostly working and these are very nitpicky "polishing" issues but in general I approve of this PR

SharkPool-SP committed:
bug fixes

SharkPool-SP commented:
@Brackets-Coder I fixed all those issues, the `There are too many active WebGL contexts on this page, the oldest context will be lost` bug I fixed in my previous commit, I used the same script in a forever loop, without refresh loop, and more... and couldnt get it

Brackets-Coder reviewed (commented):
- [x] Interpolation before creating an emitter has been fixed
- [x] Delete and wait lacking interpolation has been fixed



- [ ] The WebGL error still appears, I got it by spam-clicking the above stack in the running editor. The issue seems to be the following:

```javascript
    const disposeEngine = (target) => {
        const engine = target[engineTag];
        if (!engine)
            return;
        vm.off("STAGE_SIZE_CHANGED", engine.drawable.stageSZChange);
        render.destroyDrawable(engine.drawableId, "sprite");
        render.destroySkin(engine.skinId);
        runtime.requestRedraw();

        engine.gl.deleteProgram(engine.programInfo.program);
        const glExtension = engine.gl.getExtension("WEBGL_lose_context");
        // >>>> here
        if (glExtension)
            glExtension.loseContext();
        engine.canvas.remove();
        target[engineTag] = undefined;
        allEngines.delete(target.getName());
    };
```

I'm getting the "Too many contexts, last context will be lost" error before I get a 
"WebGL: INVALID_OPERATION: loseContext: context already lost"

It implies that loseContext isn't working because the browser automatically disposes of the context, which reveals a potential flaw with `remove all`. Maybe, if the argument is `all`, try simply deleting every existing WebGL context every time, so that when the block is spammed you'll never create too many?

I don't know. This seems like a non-trivial issue and something that could be looked further into. I don't know what's going on and since you can't replicate it on your machine, I don't know how to help you. 

<img width="1840" height="1196" alt="Image" src="https://github.com/user-attachments/assets/1243a847-ee24-445b-b2c7-1a6c91cf43d6" />

**I'm literally *this* 🤏 close to approval**. You can probably show a video demonstrating how this issue is not reproducible, fix it really quick (which I doubt this is an easy-fix), or do something... But if you don't know what's going on and I don't know what's going on, I can either approve this right away or we can get another moderator to review and help us out

SharkPool-SP commented:
Is that safari?

Brackets-Coder commented:
> Is that safari?

Ah... That might be what's causing the issue. I'll test in Firefox and probably chrome too if I feel like it

PPPDUD commented:
@SharkPool-SP Are you sure that this shouldn't be a draft? It looks like you all still need a fair bit of work before merging.

Brackets-Coder reviewed (approved):
That's the issue.

Working properly in Firefox, only returned `WebGL context was lost.` 93 times and no error. 

I'm going to approve since the issue on Safari really only shows up in the console and doesn't seem to affect performance

Brackets-Coder commented:
> Are you sure that this shouldn't be a draft? It looks like you all still need a fair bit of work before merging.

Based on what I've tested the most recent version is working as expected aside from the negligible issue above which only affects a small fraction of the population if they're looking in the console.

Any other issues I might've missed may be weeded out with a second review

GarboMuffin reviewed (changes requested):
Writing something. Hold on

GarboMuffin commented:
<img width="1318" height="612" alt="image" src="https://github.com/user-attachments/assets/c85eb635-a2c0-447e-bf32-e078f9ed4ae0" />

It's pixelated compared to the cat. I guess we never made this thing follow high quality pen?

If I want to make the particle engine work the same as high quality pen then what code am I expected to write?

The default transparency mode is also doing something wrong. The default particle colors are pink/purple/blue once they finish fading in, but while they're fading in, we see one here that is pure red. That doesn't make sense

Now zoom in on one of the particles. These particles are a solid purple but the edges are clearly pink. Not the same color. Probably related to the same transparency issue

<img width="303" height="368" alt="image" src="https://github.com/user-attachments/assets/537dba06-54a4-46bc-88b9-f19399b2bb48" />

There's also weird things going on with stacked transparent particles. These two particles overlapping should not form a deep red

<img width="508" height="381" alt="image" src="https://github.com/user-attachments/assets/2eeb302e-af8c-4090-8f82-4c05a5b8ef59" />

Brackets-Coder commented:
> I guess we never made this thing follow high quality pen?

@GarboMuffin see https://github.com/TurboWarp/extensions/pull/1724#discussion_r2689611422

GarboMuffin commented:
Here's a concrete example of a simple project using this extension

[Untitled-48.zip](https://github.com/user-attachments/files/24849731/Untitled-48.zip) (sb3 inside the zip)

What blocks will make the particles look good at any resolution?

---

Stopping project maybe should delete all the emitters. Currently if you use dynamic emitter names at all, if you want to stop seeing a screen full of particles your easiest way out is to manually run the delete engine block on whatever sprite. Which seems silly.

---

This throws an uncaught error

<img width="1008" height="300" alt="image" src="https://github.com/user-attachments/assets/898e46f7-2a9f-4e92-9a67-2d712214cb49" />

PPPDUD commented:
See what I mean?

SharkPool-SP commented:
> > Is that safari?
> 
> 
> 
> Ah... That might be what's causing the issue. I'll test in Firefox and probably chrome too if I feel like it

Might be the cause, safari is garbage

SharkPool-SP commented:
> @SharkPool-SP Are you sure that this shouldn't be a draft? It looks like you all still need a fair bit of work before merging.

I mean not really anymore, at the time I pull requested it, it was ready. Now it's just minor changes

SharkPool-SP committed:
Particle Engine 2.0.5

- framebuffers
- bug fixes
- better canvas quality

SharkPool-SP committed:
Update Particle-Engine.md

SharkPool-SP committed:
unused variable

SharkPool-SP commented:
!format

DangoCat committed:
[Automated] Format code

SharkPool-SP commented:
Okay I fixed all the issues and more
